File: benchmarks/xorfiles.rb
#!/usr/bin/env ruby

if ARGV.length != 2 then
  puts "Usage: ./xorfiles FILE1 FILE2"
  exit
end

file1, file2, *_ = ARGV

# Get file descriptors for the files
f1desc = IO.sysopen(file1)
f2desc = IO.sysopen(file2)

# Open IO streams for the files
f1 = IO.new(f1desc, 'r')
f2 = IO.new(f2desc, 'r')

BUFSIZE = 2048

begin
  while true do
    buf2 = f2.readpartial(BUFSIZE).bytes
    print f1.readpartial([buf2.length, BUFSIZE].min).bytes
      .zip(buf2)
      .map { |x, y| x ^ y }
      .map(&:chr)
      .join("")
  end
rescue EOFError
end

# Close the file descriptors
f1.close
f2.close

File: benchmarks/xorfiles.py
#!/usr/bin/env python

import sys

if len(sys.argv) != 3:
    print("Usage: ./xorfiles.py FILE1 FILE2")
    sys.exit()

_, file1, file2, *_ = sys.argv

BUFSIZE = 4096

with open(file1, 'rb') as f1:
    with open(file2, 'rb') as f2:
        while True:
            buf1 = f1.read(BUFSIZE)
            buf2 = f2.read(BUFSIZE)
            # When .read() runs out of file it returns an empty
            # string instead of, you know, erroring or otherwise
            # indicating that the file has run out -.-
            if buf1 == b'' or buf2 == b'':
                break
            print(''.join(map(
                lambda x: chr(x[0] ^ x[1]),
                zip(buf1, buf2)
                )), end='')
print("")
